### PR TITLE
INTEL MKL: Quantized Concat

### PR DESCRIPTION
**Pull request by mahmoud-abuzaina**
This PR adds support for MKL Quantized Concat op.

### CODE REVIEW

File: tensorflow/core/kernels/mkl_concat_op.cc [hunk 1]
```diff
@@ -226,9 +226,53 @@ class MklConcatOp : public OpKernel {
...
+        OP_REQUIRES_OK(context,
+                       context->input_list("input_maxes", &input_maxes));
+        OP_REQUIRES(context, (input_maxes.size() == N),
+                    errors::InvalidArgument(
+                        "QuantizedConcatOp : Expected maxes input list length ",
+                        input_maxes.size(), " to equal values length ", N));
+        float input_min = input_mins[0].flat<float>()(0);
+        float input_max = input_maxes[0].flat<float>()(0);
+        const float eps = 1.0e-6;
+        for (int i = 1; i < N; i++) {
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_concat_op.cc]:
```suggestion
        for (int i = 1; i < N; ++i) {
```

File: tensorflow/core/kernels/mkl_quantized_concat_op_test.cc [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+}
+
+void QuantizedConcatTest::TestSmall8Bit(float first_min, float first_max,
+                                        float second_min, float second_max) {
+  TF_ASSERT_OK(NodeDefBuilder("quantized_concat_op", "_MklQuantizedConcatV2")
+                   .Input(FakeInput(2, DT_QUINT8))
+                   .Input(FakeInput(DT_INT32))
+                   .Input(FakeInput(2, DT_FLOAT))
+                   .Input(FakeInput(2, DT_FLOAT))
+                   .Input(FakeInput(2, DT_UINT8))  // MKl second tensor
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_quantized_concat_op_test.cc]:
```suggestion
                   .Input(FakeInput(2, DT_UINT8))  // MKL second tensor
```
Please fix in other places as well.

File: tensorflow/core/kernels/mkl_quantized_concat_op_test.cc [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+  AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+  AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+  AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+  TF_ASSERT_OK(RunOpKernel());
+  const Tensor& output_quantized = *GetOutput(0);
+  const float output_min = GetOutput(1)->flat<float>()(0);
+  const float output_max = GetOutput(2)->flat<float>()(0);
+  Tensor output_float =
+      QuantizedTensorToFloat<quint8>(output_quantized, output_min, output_max);
+  test::ExpectTensorNear<float>(expected_float, output_float, 1.0);
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_quantized_concat_op_test.cc]: Is the error that big? I'm curious.

File: tensorflow/core/ops/mkl_array_ops.cc [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      for (int i = n + 1; i < c->num_inputs() / 2; ++i) {
+        TF_RETURN_IF_ERROR(c->WithRank(c->input(i), 0, &unused));
+      }
+      c->set_output(1, c->Scalar());
+      c->set_output(2, c->Scalar());
+      return Status::OK();
+    });
+}
+
+#endif
```
Comment (penpornk) [inline tensorflow/core/ops/mkl_array_ops.cc]: Please add a newline at the end of file.

File: tensorflow/core/ops/mkl_array_ops.cc [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+#include "tensorflow/core/util/tensor_format.h"
+
+namespace tensorflow {
+
+using shape_inference::DimensionHandle;
+using shape_inference::InferenceContext;
+using shape_inference::ShapeHandle;
+using shape_inference::UnchangedShape;
+
+// Adding QuantizedConcatV2 which is similar to the existing QuantizedConcat
```
Comment (penpornk) [inline tensorflow/core/ops/mkl_array_ops.cc]: If it's similar to `QuantizedConcat`, why don't you use `QuantizedConcat`?

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR! I have minor comments.

File: tensorflow/core/kernels/mkl_quantized_concat_op_test.cc [hunk 6]
```diff
@@ -0,0 +1,229 @@
+/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_quantized_concat_op_test.cc]:
```suggestion
/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
```

File: tensorflow/core/ops/mkl_array_ops.cc [hunk 7]
```diff
@@ -0,0 +1,92 @@
+/* Copyright 2015 The TensorFlow Authors. All Rights Reserved.
```
Comment (penpornk) [inline tensorflow/core/ops/mkl_array_ops.cc]:
```suggestion
/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
```

File: tensorflow/core/kernels/mkl_concat_op.cc [hunk 8]
```diff
@@ -226,9 +226,53 @@ class MklConcatOp : public OpKernel {
       // format and avoid calling eigen version.
       if (!are_all_tf_inputs && !are_all_mkl_inputs) invoke_eigen = true;
 
+      OpInputList input_mins, input_maxes;
+      if (std::is_same<T, qint8>::value || std::is_same<T, quint8>::value) {
+        // MKL DNN concat does not support input tensors that have different
+        // ranges, check if the ranges of the all input tensors are the same
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl_concat_op.cc]: Done.

File: tensorflow/core/kernels/mkl_concat_op.cc [hunk 9]
```diff
@@ -226,9 +226,53 @@ class MklConcatOp : public OpKernel {
...
+        // if not, forward it to Eigen implementation.
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl_concat_op.cc]: Done.

File: tensorflow/core/kernels/mkl_concat_op.cc [hunk 10]
```diff
@@ -226,9 +226,53 @@ class MklConcatOp : public OpKernel {
       // format and avoid calling eigen version.
       if (!are_all_tf_inputs && !are_all_mkl_inputs) invoke_eigen = true;
 
+      OpInputList input_mins, input_maxes;
+      if (std::is_same<T, qint8>::value || std::is_same<T, quint8>::value) {
+        // MKL DNN concat does not support input tensors that have different
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl_concat_op.cc]: Done.

File: tensorflow/core/kernels/mkl_concat_op.cc [hunk 11]
```diff
@@ -226,9 +226,53 @@ class MklConcatOp : public OpKernel {
...
+            break;
+          }
+        }
+      }
+
       // Call Eigen library
       if (invoke_eigen) {
-        CallEigenVersion(context, input_tensors, mkl_input_shapes);
+        if (std::is_same<T, qint8>::value || std::is_same<T, quint8>::value) {
+          // MKL DNN quantized concat does not support input tensors with
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl_concat_op.cc]: Done.

File: tensorflow/core/kernels/mkl_concat_op.cc [hunk 12]
```diff
@@ -226,9 +226,53 @@ class MklConcatOp : public OpKernel {
...
+          // different ranges.
+          // TODO (mabuzain): Add quantized version of CallEigen() to support
+          // this case.
+          OP_REQUIRES(context, false,
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl_concat_op.cc]: Done.

File: tensorflow/core/kernels/mkl_quantized_concat_op_test.cc [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/lib/core/status_test_util.h"
+#include "tensorflow/core/platform/test.h"
+#include "tensorflow/core/platform/test_benchmark.h"
+
+namespace tensorflow {
+
+using test::graph::Constant;
+
+// Helper class for converting MKL tesnors to TF tensors and comparing to
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl_quantized_concat_op_test.cc]: Done.

File: tensorflow/core/kernels/mkl_concat_op.cc (hunk 1 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl_concat_op.cc]: Done.

File: tensorflow/core/kernels/mkl_quantized_concat_op_test.cc (hunk 2 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl_quantized_concat_op_test.cc]: Done.

File: tensorflow/core/kernels/mkl_quantized_concat_op_test.cc (hunk 3 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl_quantized_concat_op_test.cc]: Actually, here we used the same tolerance that Eigen uses for the similar concat test.

File: tensorflow/core/ops/mkl_array_ops.cc (hunk 4 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/ops/mkl_array_ops.cc]: Done.

File: tensorflow/core/ops/mkl_array_ops.cc (hunk 5 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/ops/mkl_array_ops.cc]: Actually my comment was not clear, I fixed it. We support MKLConcatV2 for the FP32 and we use the same kernel to support MKLQuanitzedConcatV2.

File: tensorflow/core/kernels/mkl_quantized_concat_op_test.cc (hunk 6 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl_quantized_concat_op_test.cc]: Done.

File: tensorflow/core/ops/mkl_array_ops.cc (hunk 7 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/ops/mkl_array_ops.cc]: Done.

File: tensorflow/core/kernels/mkl_quantized_concat_op_test.cc (hunk 3 above)
Comment (penpornk) [reply tensorflow/core/kernels/mkl_quantized_concat_op_test.cc]: Got it. Could you please explain that in the code as a comment? Thank you!

Comment (penpornk) [reply tensorflow/core/kernels/mkl_quantized_concat_op_test.cc]: Got it. Could you please explain that in the code as a comment? Thank you!

**Review by penpornk — CHANGES_REQUESTED** [review]
One more change, please!
(I originally posted that as a comment, but thought this is easier to keep track.)

File: tensorflow/core/kernels/mkl_quantized_concat_op_test.cc (hunk 3 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl_quantized_concat_op_test.cc]: Done.

**Comment by mahmoud-abuzaina** [comment]
@penpornk Sure, no problem. I added a comment. Please let me know if anything else is needed.

**Review by penpornk — APPROVED** [review]
Great! Thank you again for the PR! :)

**Comment by penpornk** [comment]
This PR was rolled back because of test failures. I'll put the changes back once it doesn't cause anymore failures. In the meanwhile, please let me know if there are other PRs that depends on this and shouldn't be merged before this PR's changes are back.

**Comment by mahmoud-abuzaina** [comment]
@penpornk Do you mean that we need to fix something in the PR so it does not cause test failures?

**Comment by penpornk** [comment]
@mahmoud-abuzaina I already have the fix. The problem came from registering `QuantizedConcatV2` in `quantized_concat_op.cc` without the `#ifdef INTEL_MKL` guard. So non config=mkl builds try to find this op but couldn't find it. 

Now I'm debating whether I should just re-submit the changes internally (but then it wouldn't show that you authored them anymore) or ask you to open a new PR. What do you think?

**Comment by penpornk** [comment]
The new PR route should also be quick. After you re-open a new PR, please mention me and I can re-review it quickly. Please let me know what way you prefer. :)

**Comment by mahmoud-abuzaina** [comment]
Thank you @penpornk for the explanation. I will create a new PR now and mention you.

**Comment by penpornk** [comment]
@mahmoud-abuzaina Thank you and sorry for the inconvenience!

**Comment by mahmoud-abuzaina** [comment]
@penpornk I have created the replacement PR #25533. Thank you for your patience.

**Pull request merged**